src/test/QuestionAPI.t.sol

```diff
@@ -123,8 +123,39 @@ contract QuestionAPITest is Test {
         uint256 questionId = _questionAPI.createQuestion("ipfs://XYZ", 25);
```

Comment:
can we think maybe to restructure these tests - createQuestion is a question that writes state and will only be called externally I believe - but like I mentioned in the slack chat you can return values for function that write state between contracts but when you call externally it will return the value in a hash after transaction is written to blockchain, so in reality questionId is not quite available as it in these tests - so my question is how much value do they have as I don't believe this function will be called internally but always through a dApp or some kind of external script

Reply:
I'm not sure that I am fully understanding. All functionality relies on the questionId -- which is just a uint. What return value is hashed? We pretty much just log out the tokenId of the minted question which we need to update/change any of the properties tied to it, whether that be the amount of votes, metric locked etc. Could you maybe send me an example of what you mean? Tyty

Reply:
when you call this function externally meaning not via internally via a solidity function it writes state to the blockchain - questionId is not readily available until it is written to the blockchain and the return value is a hash. which is not really returned so it's returned when you write it in solidity but real world use case we won't be calling this function internally or form another contract but externally like a dApp - https://ethereum.stackexchange.com/questions/88119/i-see-no-way-to-obtain-the-return-value-of-a-non-view-function-ethers-js

Reply:
For example in the js test you call something like 
```
      const questionIDtx2 = await questionAPI.connect(xmetricaddr1).createQuestion("metricsdao.xyz", 10);
      const waitforTx = await questionIDtx2.wait();
      console.log("waitforTx", waitforTx);
      ```
      It will return below object



```
waitforTx {
  to: '0x7A9Ec1d04904907De0ED7b6839CcdD59c3716AC9',
  from: '0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC',
  contractAddress: null,
  transactionIndex: 0,
  gasUsed: BigNumber { value: "316141" },
  logsBloom: '0x04000800000010000000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000300000000000000000000000000008000000000000000000000000000000000000000000000000020000000000000000000800000800000000000000002010000000000000000000000000000000000000000000000000000200000000000000000000020000000000200000000100000000000000000000000000000400000000000000000202000000010000000000000000000000080000000000000000000020000010000000000000000000000000000000000000008000000000000000000000',
  blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d',
  transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
  logs: [
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0x46b142DD1E924FAb83eCc3c08e4D46E82f005e0E',
      topics: [Array],
      data: '0x0000000000000000000000000000000000000000000000000000000000000000',
      logIndex: 0,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d'
    },
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0x46b142DD1E924FAb83eCc3c08e4D46E82f005e0E',
      topics: [Array],
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      logIndex: 1,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d'
    },
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0x46b142DD1E924FAb83eCc3c08e4D46E82f005e0E',
      topics: [Array],
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      logIndex: 2,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d'
    },
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0xC9a43158891282A2B1475592D5719c001986Aaec',
      topics: [Array],
      data: '0x',
      logIndex: 3,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d'
    }
  ],
  blockNumber: 321,
  confirmations: 1,
  cumulativeGasUsed: BigNumber { value: "316141" },
  effectiveGasPrice: BigNumber { value: "1000000007" },
  status: 1,
  type: 2,
  byzantium: true,
  events: [
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0x46b142DD1E924FAb83eCc3c08e4D46E82f005e0E',
      topics: [Array],
      data: '0x0000000000000000000000000000000000000000000000000000000000000000',
      logIndex: 0,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d',
      removeListener: [Function (anonymous)],
      getBlock: [Function (anonymous)],
      getTransaction: [Function (anonymous)],
      getTransactionReceipt: [Function (anonymous)]
    },
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0x46b142DD1E924FAb83eCc3c08e4D46E82f005e0E',
      topics: [Array],
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      logIndex: 1,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d',
      removeListener: [Function (anonymous)],
      getBlock: [Function (anonymous)],
      getTransaction: [Function (anonymous)],
      getTransactionReceipt: [Function (anonymous)]
    },
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0x46b142DD1E924FAb83eCc3c08e4D46E82f005e0E',
      topics: [Array],
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      logIndex: 2,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d',
      removeListener: [Function (anonymous)],
      getBlock: [Function (anonymous)],
      getTransaction: [Function (anonymous)],
      getTransactionReceipt: [Function (anonymous)]
    },
    {
      transactionIndex: 0,
      blockNumber: 321,
      transactionHash: '0x63d60a2e18fe77383ad185a189742bfc18eec755bb07a20c1a1c80db59b3b2c7',
      address: '0xC9a43158891282A2B1475592D5719c001986Aaec',
      topics: [Array],
      data: '0x',
      logIndex: 3,
      blockHash: '0x3c4e6726a3a2d88de0d5b8d58847b085bebd58d6c67067d64cc906036bf94c5d',
      removeListener: [Function (anonymous)],
      getBlock: [Function (anonymous)],
      getTransaction: [Function (anonymous)],
      getTransactionReceipt: [Function (anonymous)]
    }
  ]
}
```

Reply:
Thank you for explaining. Will brainstorm

Reply:
I have created a getter and setter for this - https://github.com/MetricsDAO/core-evm-contracts/pull/95/files

Functionality wise yes you are returning that data - but in a real world use case I don't see calling createQuestion from anything that isn't external and when you write data to the blockchain you typically don't rely on the return data from that write but create a setter and a getter to retrieve that data later - while we are testing if it returns the "right" data but the reality is these tests aren't a good example of the real world use case - for example my JS tests will need to be structured far differently and I'm not sure testing frameworks should diverge that much